lang/fastrand/fastrand.go

```diff
@@ -125,26 +125,81 @@ func Uint64n(n uint64) uint64 {
 }
 
 // Read generates len(p) random bytes and writes them into p.
-// It always returns len(p) and a nil error. And it is safe
-// for concurrent use.
-func Read(p []byte) (n int, err error) {
+// It always returns len(p) and a nil error.
+// It is safe for concurrent use.
+func Read(p []byte) (int, error) {
 	l := len(p)
 
+	// Used for local XORSHIFT.
+	var tmp [2]uint32
+	tmp[0], tmp[1] = Uint32(), Uint32()
+
 	if l >= 4 {
-		i := 0
+		var i int
 		uint32p := *(*[]uint32)(unsafe.Pointer(&p))
-		for ; l >= 4; l -= 4 {
-			uint32p[i] = Uint32()
+		s1, s0 := tmp[0], tmp[1]
+		for l >= 4 {
+			// Local XORSHIFT.
+			s1 ^= s1 << 17
+			s1 = s1 ^ s0 ^ s1>>7 ^ s0>>16
+			s0, s1 = s1, s0
+
+			uint32p[i] = s0 + s1
 			i++
+			l -= 4
 		}
+		tmp[0], tmp[1] = s1, s0
 	}
 
 	if l > 0 {
-		r := Uint32()
-		for ; l > 0; l-- {
+		// Local XORSHIFT.
+		// We don't need to save s0 and s1(tmp is not used).
+		s1, s0 := tmp[0], tmp[1]
+		s1 ^= s1 << 17
+		s1 = s1 ^ s0 ^ s1>>7 ^ s0>>16
+
+		r := s0 + s1
+		for l > 0 {
 			p[len(p)-l] = byte(r >> (l * 8))
+			l--
 		}
 	}
 
 	return len(p), nil
 }
+
+// Shuffle pseudo-randomizes the order of elements.
+// n is the number of elements. Shuffle panics if n < 0.
+// swap swaps the elements with indexes i and j.
+func Shuffle(n int, swap func(i, j int)) {
+	if n < 0 {
+		panic("invalid argument to Shuffle")
+	}
+	// Fisher-Yates shuffle: https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
+	// Shuffle really ought not be called with n that doesn't fit in 32 bits.
+	// Not only will it take a very long time, but with 2³¹! possible permutations,
+	// there's no way that any PRNG can have a big enough internal state to
+	// generate even a minuscule percentage of the possible permutations.
+	// Nevertheless, the right API signature accepts an int n, so handle it as best we can.
+	i := n - 1
+	for ; i > 1<<31-1-1; i-- {
```

Comment:
I was confused by `1<<32-1-1`, shall we just use `1 << 29`? or at least `1<<(31-1-1)`, hope I was not wrong about the priority :(

and we may need some explanation on `32-1-1`

Reply:
TODO